### ISSUE TITLE
IO-Tty-1.18 fails to install on Mac OS X

### ISSUE THREAD

**Issue by ferki** [cpan-authors/IO-Tty#37]
Apparently IO-Tty-1.18 fails to install on Mac OS X systems. I first noticed it as a build failure on such systems while installing dependencies of a project I maintain. [Results from CPAN Testers](http://fast-matrix.cpantesters.org/?dist=IO-Tty%201.18) seems to match that experience too.

It seems to work on other operating systems, and installing the previous version(s) seems working on Mac OS X too.

Unfortunately I do not have access to such systems apart from GitHub Action runners, nor do I have experience with IO-Tty internals itself, so I could not look into root causes so far. I'm not yet sure how I may be more helpful apart from opening this issue – please let me know.

**Comment by DrHyde** [comment]
Looks like an odd C error:

```
  Tty.xs:190:1: error: expected ')'
  strlcpy(dst, src, siz)
```

That looks perfectly cromulent, but maybe it's a funky macro expansion going wrong.

**Comment by toddr** [comment]
I didn't change much in this release, so I wonder if it's more an issue with new versions of OSX.

**Comment by grr** [comment]
The contents from conf/functest_strlcpy.log after running `perl Make.PL` looks relevant:

    functest_strlcpy.c:10:6: warning: incompatible redeclaration of library function 'strlcpy' [-Wincompatible-library-redeclaration]
    char strlcpy ();
        ^
    functest_strlcpy.c:10:6: note: 'strlcpy' is a builtin with type 'unsigned long (char *, const char *, unsigned long)'
    1 warning generated.
    ld: unknown option: --no-undefined
    clang: error: linker command failed with exit code 1 (use -v to see invocation)

And this is on an older macOS version (Big Sur - 11.7.10).

**Comment by ferki** [comment]
> I didn't change much in this release, so I wonder if it's more an issue with new versions of OSX.

Hmm, from _my perspective_ it's still only IO-Tty-1.18 that fails and only on Mac. (So I might update my project to use only IO-Tty-1.17 on Mac just to unblock regular testing for now).

The full text from the list of available [CPAN Testers reports for darwin](http://fast-matrix.cpantesters.org/?dist=IO-Tty%201.18;os=darwin;reports=1) indicate multiple different Mac OS (kernel?) versions, LLVM/clang versions, Perl versions, etc., are involved, as well as at least two different type of failures.

IO-Tty-1.17 seems to work both for CPAN Testers and GitHub Actions runners, and I couldn't identify an OS change that would indicate a failure like this :thinking: 

To me it appears as a build failure, and the last release changed the build flags. This repo's GitHub Action tests are failing on Mac OS too since the only change merged prior to the 1.18 release (ideally that could have prevented releasing 1.18).

Just to throw in some quick workaround ideas:

- Would it be a viable option to revert the last version's changes for now, make a clean 1.19 release (otherwise identical to 1.17), and make sure the proposed fix works as expected before releasing 1.20?
- Or perhaps remove 1.18 from CPAN until a proper fix for 1.19?

**Comment by DrHyde** [comment]
On this 'ere Mac, 1.17 is just fine and 1.18 isn't, with exactly the same build of perl, compiler, etc:

```
~ $ cpanm IO::Tty@1.17
--> Working on IO::Tty
Fetching http://www.cpan.org/authors/id/T/TO/TODDR/IO-Tty-1.17.tar.gz ... OK
Configuring IO-Tty-1.17 ... OK
Building and testing IO-Tty-1.17 ... OK
Successfully installed IO-Tty-1.17
1 distribution installed
~ $ cpanm IO::Tty@1.18
! Finding IO::Tty (== 1.18) on cpanmetadb failed.
--> Working on IO::Tty
Fetching http://cpan.metacpan.org/authors/id/T/TO/TODDR/IO-Tty-1.18.tar.gz ... OK
Configuring IO-Tty-1.18 ... OK
Building and testing IO-Tty-1.18 ... FAIL
! Installing IO::Tty failed. See /Users/david/.cpanm/work/1701287467.64323/build.log for details. Retry with --force to force install it.
```

**Comment by mrchulho** [comment]
A linker option was added to 1.18.
```
-  if (system("$cfg{'cc'} $flags $funcs{$f} functest_$f.c > functest_$f.log 2>&1")) {
+  if (system("$cfg{'cc'} $flags -Wl,--no-undefined $funcs{$f} functest_$f.c > functest_$f.log 2>&1")) {
```
[https://metacpan.org/release/TODDR/IO-Tty-1.18/diff/TODDR/IO-Tty-1.17#Makefile.PL](https://metacpan.org/release/TODDR/IO-Tty-1.18/diff/TODDR/IO-Tty-1.17#Makefile.PL)

Apple `ld` linker expects `-Wl,-undefined,error` instead of `-Wl,--no-undefined`. 
[https://stackoverflow.com/questions/26971333/what-is-clangs-equivalent-to-no-undefined-gcc-flag](https://stackoverflow.com/questions/26971333/what-is-clangs-equivalent-to-no-undefined-gcc-flag)

ld man page reads:

```
     -undefined treatment
             Specifies how undefined symbols are to be treated. Options are:
             error, warning, suppress, or dynamic_lookup.  The default is
             error. Note: dynamic_lookup that depends on lazy binding will not
             work with chained fixups.
```

Tested in Sonoma 14.1.1, Xcode 15.0.1
% cc --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: x86_64-apple-darwin23.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
% ld -v
@(#)PROGRAM:ld  PROJECT:dyld-1015.7
BUILD 16:59:22 Oct  1 2023
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
will use ld-classic for: armv6 armv7 armv7s arm64_32 i386 armv6m armv7k armv7m armv7em
LTO support using: LLVM version 15.0.0 (static support for 29, runtime is 29)
TAPI support using: Apple TAPI version 15.0.0 (tapi-1500.0.12.3)
Library search paths:
Framework search paths:

**Comment by rshingleton** [comment]
I'm running into the same issue when install on a new M2

OS: 14.1.2 (23B92)

**STACKTRACE:**

2023-12-12T14:33:26,14630,IO-Tty-1.18| "/Users/shingler/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Tty.bs blib/arch/auto/IO/Tty/Tty.bs 644
2023-12-12T14:33:26,14630,IO-Tty-1.18| "/Users/shingler/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/bin/perl" "/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/lib/5.36.3/ExtUtils/xsubpp"  -typemap '/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/lib/5.36.3/ExtUtils/typemap'  Tty.xs > Tty.xsc
2023-12-12T14:33:26,14630,IO-Tty-1.18| mv Tty.xsc Tty.c
2023-12-12T14:33:26,14630,IO-Tty-1.18| cc -c   -fPIC -fno-strict-aliasing -pipe -fstack-protector-strong -DPERL_USE_SAFE_PUTENV -Wno-error=implicit-function-declaration -O3   -DVERSION=\"1.18\" -DXS_VERSION=\"1.18\"  "-I/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/lib/5.36.3/darwin-thread-multi-2level/CORE"  -DHAVE_DEV_PTMX -DHAVE_TERMIOS_H -DHAVE_UTIL_H -DUTIL_H_ABS_PATH=\"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/util.h\" Tty.c
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: error: expected ')'
2023-12-12T14:33:26,14630,IO-Tty-1.18| strlcpy(dst, src, siz)
2023-12-12T14:33:26,14630,IO-Tty-1.18| ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:47: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                             ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_common.h:39:53: note: expanded from macro '__darwin_obsz'
2023-12-12T14:33:26,14630,IO-Tty-1.18| #define __darwin_obsz(object) __builtin_object_size (object, _USE_FORTIFY_LEVEL > 1 ? 1 : 0)
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                     ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: note: to match this '('
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:27: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                         ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: warning: parameter '__builtin_object_size' was not declared, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
2023-12-12T14:33:26,14630,IO-Tty-1.18| strlcpy(dst, src, siz)
2023-12-12T14:33:26,14630,IO-Tty-1.18| ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:47: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                             ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_common.h:39:31: note: expanded from macro '__darwin_obsz'
2023-12-12T14:33:26,14630,IO-Tty-1.18| #define __darwin_obsz(object) __builtin_object_size (object, _USE_FORTIFY_LEVEL > 1 ? 1 : 0)
2023-12-12T14:33:26,14630,IO-Tty-1.18|                               ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:3: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| 2 warnings and 1 error generated.
2023-12-12T14:33:26,14630,IO-Tty-1.18| make: *** [Tty.o] Error 1
2023-12-12T14:33:26,14630,IO-Tty-1.18| ! Retrying (you can turn off this behavior by --no-retry)
2023-12-12T14:33:26,14630,IO-Tty-1.18| Executing /usr/bin/make
2023-12-12T14:33:26,14630,IO-Tty-1.18| "/Users/shingler/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Tty.bs blib/arch/auto/IO/Tty/Tty.bs 644
2023-12-12T14:33:26,14630,IO-Tty-1.18| cc -c   -fPIC -fno-strict-aliasing -pipe -fstack-protector-strong -DPERL_USE_SAFE_PUTENV -Wno-error=implicit-function-declaration -O3   -DVERSION=\"1.18\" -DXS_VERSION=\"1.18\"  "-I/prod/dependencies/perl5/perlbrew/perls/perl-5.36.3/lib/5.36.3/darwin-thread-multi-2level/CORE"  -DHAVE_DEV_PTMX -DHAVE_TERMIOS_H -DHAVE_UTIL_H -DUTIL_H_ABS_PATH=\"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/util.h\" Tty.c
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: error: expected ')'
2023-12-12T14:33:26,14630,IO-Tty-1.18| strlcpy(dst, src, siz)
2023-12-12T14:33:26,14630,IO-Tty-1.18| ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:47: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                             ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_common.h:39:53: note: expanded from macro '__darwin_obsz'
2023-12-12T14:33:26,14630,IO-Tty-1.18| #define __darwin_obsz(object) __builtin_object_size (object, _USE_FORTIFY_LEVEL > 1 ? 1 : 0)
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                     ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: note: to match this '('
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:27: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                         ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: warning: parameter '__builtin_object_size' was not declared, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
2023-12-12T14:33:26,14630,IO-Tty-1.18| strlcpy(dst, src, siz)
2023-12-12T14:33:26,14630,IO-Tty-1.18| ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:47: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                                                             ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_common.h:39:31: note: expanded from macro '__darwin_obsz'
2023-12-12T14:33:26,14630,IO-Tty-1.18| #define __darwin_obsz(object) __builtin_object_size (object, _USE_FORTIFY_LEVEL > 1 ? 1 : 0)
2023-12-12T14:33:26,14630,IO-Tty-1.18|                               ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| Tty.xs:190:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
2023-12-12T14:33:26,14630,IO-Tty-1.18| /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_string.h:112:3: note: expanded from macro 'strlcpy'
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 __builtin___strlcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
2023-12-12T14:33:26,14630,IO-Tty-1.18|                 ^
2023-12-12T14:33:26,14630,IO-Tty-1.18| 2 warnings and 1 error generated.
2023-12-12T14:33:26,14630,IO-Tty-1.18| make: *** [Tty.o] Error 1
2023-12-12T14:33:27,14630,IO-Tty-1.18| Failed to install distribution
2023-12-12T14:33:27,14613| --
2023-12-12T14:33:27,14613| Installation failed. The direct cause of the failure comes from the following packages/distributions; you may want to grep this log file by them:
2023-12-12T14:33:27,14613|  * IO-Tty-1.18


Seems to be new in 1.18, was able to get IO::Tty@1.17 to install without issues:
cpm install -g --show-build-log-on-failure  IO::Tty@1.17
DONE install IO-Tty-1.17
1 distribution installed.

**Comment by abqcheeks** [comment]
FreeBSD has a similar issue.  The FreeBSD port ships with patch for Tty.xs which changes the definition of strlcpy() from "static size_t" to "size_t".  This allows it to compile, but it doesn't actually work after installation (can't open any ptys.  I'll open a separate issue for the FreeBSD situation, but I Think they are related and might have the same root cause.
